cdc/log.cc

```diff
@@ -539,12 +539,109 @@ class transformer final {
         }
     }
 
+    template<typename Func>
+    static void map_for_each(const bytes_opt& bo, Func&& f) {
+        if (bo) {
+            bytes_view bv(*bo);
+            auto size = read_collection_size(bv, cql_serialization_format::internal());
+            while (size--) {
+                auto k = read_collection_value(bv, cql_serialization_format::internal());
+                auto v = read_collection_value(bv, cql_serialization_format::internal());
+                f(k, v);
+            }
+        }
+    }
+    template<typename Func>
+    static void set_for_each(const bytes_opt& bo, Func&& f) {
+        if (bo) {
+            bytes_view bv(*bo);
+            auto size = read_collection_size(bv, cql_serialization_format::internal());
+            while (size--) {
+                f(read_collection_value(bv, cql_serialization_format::internal()));
+            }
+        }
+    }
+    template<typename Func>
+    static void udt_for_each(const bytes_opt& bo, Func&& f) {
+        if (bo) {
+            bytes_view bv(*bo);
+            std::for_each(tuple_deserializing_iterator::start(bv), tuple_deserializing_iterator::finish(bv), std::move(f));
+        }
+    }
+    static bytes merge_maps(const abstract_type& type, const bytes_opt& prev, const bytes_opt& next, const bytes_opt& deleted) {
+        std::vector<std::pair<bytes_view, bytes_view>> res;
+        map_for_each(prev, [&res](bytes_view k, bytes_view v) {
+            res.emplace_back(k, v);
+        });
+        map_for_each(next, [&res, &type](bytes_view k, bytes_view v) {
+            auto i = std::find_if(res.begin(), res.end(), [&k, &type](std::pair<bytes_view, bytes_view>& p) {
+                return type.compare(p.first, k) >= 0;
+            });
```

Comment:
This is a quadratic algorithm, which means it will be unnecessarily slow even for quite small maps (I assume that 10000 keys is "small").
I'd suggest using a `map` instead of `vector` to construct the result so we get `O(n logn)` complexity.

Same comment applies to `merge_sets` below.

Reply:
I disagree. Using maps will create a plethora of nodes for every element, causing massive memory fragmentation. Since the objects are small, potential move-inserts should be relatively cheap, and actual collisions should be rate - if we overwrite, this is just a series of emplace_back, if we set an element, it is most likely only a few elements.

Reply:
> potential move-inserts should be relatively cheap

It doesn't matter when you have quadratic complexity. It is easy to create an example where a quadratic algorithm will create a 1 second stall, a 1 day stall, a 1 week stall etc.

Reply:
For ~ `10^5` elements you'll have ~ `10^10` operations, which will already take about a second if I remember correctly (from my competitive programming days I remember that a quadratic solution wouldn't pass even for input size of `10^4` :P)

Reply:
I agree with Kamil and seeing his wise words I hope he won't dispute with https://github.com/scylladb/scylla/pull/5991#discussion_r390929962 :D

Reply:
BTW @elcallio I see your point too and I'm not a great fan of maps either. Sometimes the solution is to sort the collections and then base the efficient solution on that.

Reply:
Oh yes, this can be made in a cache-and-memory-friendly way using a [merge algorithm](https://en.wikipedia.org/wiki/Merge_algorithm). I think there's even a standard library function for that.

Btw, recommended read:
https://news.ycombinator.com/item?id=21743424

Reply:
That is the point. The vector is sorted. A better option is to already switch to use lower_bound, i.e. binary search.

Reply:
`lower_bound` would fix the quadraticity introduced by calling `find_if` here I agree, but there's still the problem with `emplace`s

Reply:
emplaces can be fixed with reserve.

Reply:
No, it won't fix the issue.
It will *maybe* cause allocations not to happen, but `emplace` will still be linear (it shifts elements to the right), so the entire algorithm will be quadratic.